clients/drcachesim/tests/analysis_unit_tests.cpp

```diff
@@ -0,0 +1,189 @@
+            worker_data_.push_back(analyzer_worker_data_t(i, scheduler_.get_stream(i)));
+        }
+    }
+};
+
+bool
+test_queries()
+{
+    std::cerr << "\n----------------\nTesting queries\n";
+    std::vector<trace_entry_t> input_sequence = {
+        make_thread(/*tid=*/1),
+        make_pid(/*pid=*/1),
+        make_instr(/*pc=*/42),
+        make_exit(/*tid=*/1),
+    };
+    static constexpr int NUM_INPUTS = 3;
+    static constexpr int NUM_OUTPUTS = 2;
+    static constexpr int BASE_TID = 100;
+    std::vector<trace_entry_t> inputs[NUM_INPUTS];
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    for (int i = 0; i < NUM_INPUTS; i++) {
+        memref_tid_t tid = BASE_TID + i;
+        inputs[i] = input_sequence;
+        for (auto &record : inputs[i]) {
+            if (record.type == TRACE_TYPE_THREAD || record.type == TRACE_TYPE_THREAD_EXIT)
+                record.addr = static_cast<addr_t>(tid);
+        }
+        std::vector<scheduler_t::input_reader_t> readers;
+        readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(inputs[i])),
+                             std::unique_ptr<mock_reader_t>(new mock_reader_t()), tid);
+        sched_inputs.emplace_back(std::move(readers));
+    }
+    scheduler_t::scheduler_options_t sched_ops(
+        scheduler_t::MAP_TO_ANY_OUTPUT, scheduler_t::DEPENDENCY_IGNORE,
+        scheduler_t::SCHEDULER_DEFAULTS, /*verbosity=*/3);
+
+    class test_tool_t : public analysis_tool_t {
+    public:
+        bool
+        process_memref(const memref_t &memref) override
+        {
+            assert(false); // Only expect parallel mode.
+            return false;
+        }
+        bool
+        print_results() override
+        {
+            return true;
+        }
+        bool
+        parallel_shard_supported() override
+        {
+            return true;
+        }
+        void *
+        parallel_shard_init_stream(int shard_index, void *worker_data,
+                                   memtrace_stream_t *stream) override
+        {
+            auto per_shard = new per_shard_t;
+            per_shard->index = shard_index;
+            per_shard->stream = stream;
+            return reinterpret_cast<void *>(per_shard);
+        }
+        bool
+        parallel_shard_exit(void *shard_data) override
+        {
+            per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
+            delete shard;
+            return true;
+        }
+        bool
+        parallel_shard_memref(void *shard_data, const memref_t &memref) override
+        {
+            per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
+            // These are our testing goals: these queries.
+            // We have one thread for each of our NUM_INPUTS workloads.
+            assert(shard->stream->get_output_cpuid() == shard->index);
+            assert(shard->stream->get_workload_id() == memref.instr.tid - BASE_TID);
```

Comment:
Workload id is same as the input id?

Reply:
Oh that's because the way we've set up this test: each thread is its own workload.

Add a comment.

Reply:
Done.